### DISCUSSION TITLE
[Archive] Create a separate file for each channel

### DISCUSSION THREAD

**Discussion by freearhey** [iptv-org/community#518] [Ideas]
### Idea

**Not possible to implement.**

The idea is to split the country playlists into separate files that will contain links to all the stream variants for a particular channel.

This should solve several problems:

- ~~it make possible to store a full description of the channel even if there is no working link~~ 
  This is already possible through the use of `[Offline]` status.
- ~~there will be no need to duplicate channel descriptions because each channel will have only one main link~~
  It is not possible to have links to broadcasts from different sources in one playlist, as it leads to freezes when the player automatically changes the quality of the broadcast. So this won't work:

```
#EXTM3U
#EXT-X-STREAM-INF:BANDWIDTH=7594400,RESOLUTION=1920x1080
https://rt-news.secure.footprint.net/1103_4500Kb.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=2291203,RESOLUTION=852x480
https://strm.yandex.ru/kal/rt_hd/rt_hd0_169_480p.json/index-v1-a1.m3u8
```
- ~~this will also remove all unnecessary links leading to the same version of the stream~~
  This can also be achieved by leaving only adaptive links in the playlist.

### New file structure

In the new system, each country will be represented by a folder inside which there will be `_index.m3u` file and a separate file for each channel (the file name in this case is a [kebab-case](https://textedit.tools/kebabcase) version of the channel name).

<img width="393" alt="iptv-structior-v2" src="https://user-images.githubusercontent.com/7253922/127375973-2d51faff-eab4-4fb8-84ce-bf6828b922cc.png">

**_index.m3u**

The `_index.m3u` file is essentially a normal country playlist, only the links in it lead not to third-party resources but to files within the same folder.

```
#EXTM3U
#EXTINF:-1 tvg-id="ArianaNews.af" tvg-name="Ariana News" tvg-country="AF" tvg-language="Pushto" tvg-logo="https://i.imgur.com/x8mlbJM.jpg" group-title="",Ariana News
ariana-news.m3u8
#EXTINF:-1 tvg-id="ArianaTVNational.af" tvg-name="Ariana TV National" tvg-country="AF" tvg-language="Pushto" tvg-logo="https://i.imgur.com/8VbuI9n.png" group-title="",Ariana TV National
ariana-tv-national.m3u8
#EXTINF:-1 tvg-id="ArianaTVUS.af" tvg-name="Ariana TV US" tvg-country="AF" tvg-language="Pushto" tvg-logo="https://i.imgur.com/8VbuI9n.png" group-title="",Ariana TV US
ariana-tv-us.m3u8
#EXTINF:-1 tvg-id="KayhanTV.af" tvg-name="Kayhan TV" tvg-country="AF" tvg-language="Pushto" tvg-logo="https://i.imgur.com/XpR1VvZ.png" group-title="",Kayhan TV
kayhan-tv.m3u8
#EXTINF:-1 tvg-id="PamirTV.af" tvg-name="Pamir TV" tvg-country="AF" tvg-language="Pushto" tvg-logo="" group-title="",Pamir TV
pamir-tv.m3u8
#EXTINF:-1 tvg-id="Tuti.af" tvg-name="Tuti" tvg-country="AF" tvg-language="Pushto" tvg-logo="https://i.imgur.com/JerfIBt.jpg" group-title="",Tuti
tuti.m3u8
```

The `#EXTVLCOPT` tag can be used as before:

```
#EXTM3U
#EXTINF:-1 tvg-id="21TV.am" tvg-name="21 TV" tvg-country="AM" tvg-language="Armenian" tvg-logo="https://i.imgur.com/hbkWdwn.png" group-title="",21 TV (480p)
#EXTVLCOPT:http-user-agent=Mozilla/5.0 (iPhone; CPU iPhone OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148
21-tv.m3u8
```

**ariana-news.m3u8**

This file contains links to all variants of broadcasts of the given channel. 

```
#EXTM3U
#EXT-X-STREAM-INF:BANDWIDTH=2321424,RESOLUTION=1280x720
https://d10rltuy0iweup.cloudfront.net/ATNNEWS/myStream/chunklist_w477111583.m3u8
```

If the channel has several versions of the broadcast, they can all be listed here, too:

```
#EXTM3U
#EXT-X-STREAM-INF:BANDWIDTH=125000,RESOLUTION=320x180
https://admdn5.cdn.mangomolo.com/drama/smil:drama.stream.smil/chunklist_b125000_t64MTgwcA==.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=200000,RESOLUTION=416x234
https://admdn5.cdn.mangomolo.com/drama/smil:drama.stream.smil/chunklist_b200000_t64MjM0cA==.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=450000,RESOLUTION=640x360
https://admdn5.cdn.mangomolo.com/drama/smil:drama.stream.smil/chunklist_b450000_t64MzYwcA==.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=800000,RESOLUTION=854x480
https://admdn5.cdn.mangomolo.com/drama/smil:drama.stream.smil/chunklist_b800000_t64NDgwcA==.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=1800000,RESOLUTION=1280x720
https://admdn5.cdn.mangomolo.com/drama/smil:drama.stream.smil/chunklist_b1800000_t64NzIwcA==.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=4000000,RESOLUTION=1920x1080
https://admdn5.cdn.mangomolo.com/drama/smil:drama.stream.smil/chunklist_b4000000_t64MTA4MHA=.m3u8
```

For proper work of the playlist in most players you only need to specify in it the maximum bitrate of broadcast (BANDWIDTH) and its RESOLUTION. This information can be found either by opening the channel playlist in text editor or through the "Media information..." window in VLC.

More information about this playlist format and other available attributes can be found here:

- https://developer.apple.com/documentation/http_live_streaming/example_playlists_for_http_live_streaming/creating_a_primary_playlist#overview
- https://datatracker.ietf.org/doc/html/rfc8216#section-4.3.4.2

### Transition

To avoid interrupting playlist updates for a long time, the transition to the new structure can be broken into two steps:

1. Create `_index.m3u` files and transfer all links from country playlists to them. At the same time, we will need to update all the scripts, in order not to interrupt the daily updates.

2. Gradually creating files for each channel and moving all links to broadcasts into them.

### Feedback

If you find any flaws in this plan, please let me know. Since everything is still at the stage of idea now it will be much easier to make some changes or stop its implementation altogether.

**Comment by Sphinxroot** [comment]
with this option could be usefull for  dailymotion  stream in the same format  , but those token  need to be update quite a lot , so the link checker will delete them  , but  if a  exeption tag is allowed  like [Not 24/7] .. do a script can catch the dead link without mandatory deleting them ?

those are my questioning
this will add a lot of PR for each channel posted other than 1 PR per list
do it change  the behavior of #EXTVLCOPT:  user agent and referer 
how about .mpd stream 
how about  stream that don`t have extension file 
do we only put there , stream that have Multiple RESOLUTION , because some  have a unique one
do this method will reduce the people of willing to make a PR  and share with us

**Reply by freearhey** [reply]
1. We could try add our own attribute to the stream description and tell the scripts to skip such links during checks, for example:

<code>#EXT-X-STREAM-INF:BANDWIDTH=2321424,RESOLUTION=1280x720,<b>STATUS="offline"</b></code>

2. In fact, it is not necessary for all channels to have their own files in order for playlists to work properly, although this is certainly a goal. Everything should work even if some of the links will lead as before to third-party resources.

```xml
#EXTM3U
#EXTINF:-1,Abu Dhabi Sports 3
abu-dhabi-sports-3.m3u8
#EXTINF:-1,Abu Dhabi Sports 4
https://admdn4ta.cdn.mgmlcdn.com/adsports4/smil:adsports4.stream.smil/playlist.m3u8
```

So, to quickly add channels, you will only need to update the file `<country>/_index.m3u`.

3. On the subject of the `#EXTVLCOPT` tag, that's a great question. I don't have an answer to that yet, but I'll look into it and post more later.
4. We can leave the links to `.mpd` files in the playlists as they are now, unless of course there is some other option.
5. Could you give some example of such a stream?
6. Even if the channel is only available in one resolution, it is better to create file for it in case an alternative appears in the future.
7. Honestly, I don't know. But as I wrote above, if contributors do not want to bother with creating files, they can edit only one file "_index.m3u" as it happens now.

**Reply by freearhey** [reply]
I added an example with the `#EXTVLCOPT` tag.

**Reply by tarrazuapps** [reply]
Is part of the standard for the player to recognize relative file paths?

**Reply by freearhey** [reply]
The thing is that m3u playlists have no official specification (https://en.wikipedia.org/wiki/M3U#File_format) and each developer on his own decides which features will be supported by his player. But in any case, as far as I know most if not all players support relative paths.
